[ENH] New widget: Import Documents

jerneju opened the pull request:
##### Issue
New widget. Fixes #101 


##### Description of changes
Text can now be imported from:
- Ordinary text file (`.txt`)
- Office Open XML (`.docx`)
- OpenDocument (`.odt`)
- Portable Document Format (`.pdf`)
- Extensible Markup Language (XML) (`.xml`)

New requirements:
- pdfminer3k>=1.3.1
- odfpy>=1.3.5 (https://anaconda.org/conda-forge/odfpy)
- docx2txt>=0.6 (https://anaconda.org/conda-forge/docx2txt)
- lxml

If there are more folders, folder names are also included in Corpus file. They are represented as a class variable and under name category.


##### Includes
- [X] Code changes
- [ ] Tests
- [ ] Documentation

kernc commented:
What about .doc, .odt and .pdf files? Readers surely exist?

jerneju commented:
@kernc: Hopefully yes. Check this out: http://textract.readthedocs.io/en/latest/

kernc commented on orangecontrib/text/widgets/owimporttexts.py:
```diff
@@ -0,0 +1,776 @@
+        for path in filescanner:
+            if fnmatch_any(path, patterns):
+                reader = Reader.get_reader(path)
+                textdata = reader.read()
```
I suppose this may fail with arbitrary error, e.g. on malformed PDF.

jerneju replied:
Sure, `try except` block will be added.

kernc commented:
Wouldn't it be reasonable to expose from-directory corpus loading in the scripting API as well?

jerneju commented on orangecontrib/text/widgets/owimporttexts.py:
```diff
@@ -0,0 +1,776 @@
+            self.recent_cb.setCurrentIndex(-1)
+        self.__actions.reload.setEnabled(self.currentPath is not None)
+
+    def customEvent(self, event):
```
This widget is a modified copy of **Import Images** which is a part of **Image Analytics** add-on. I need to check this.

kernc replied:
IIUC this custom event handler, as used, prevents running the `start()` method's code as part of widget constructor. Like if `QTimer.singleShot(1, ...)` were used.

ajdapretnar commented:
I understand this is WIP, but just a note while you're working: some pdfs get rendered nicely, others poorly and some not at all. Please check the reader and how/whether is works. Also there might be some issues with non-ASCII characters, please check and test this as well. Thanks! :)

Oh also:
1. remove 'group' field. Use it for _class_ only if there's more than one folder.
2. remove 'ext' field. Instead, remove the extension from filename to make it nicer to read: file.txt --> file
3. set 'content' to be a text feature. This is urgent, since without it nothing will work.

ajdapretnar commented on orangecontrib/text/widgets/owimporttexts.py:
```diff
@@ -0,0 +1,643 @@
+from Orange.canvas.preview.previewbrowser import TextLabel
+
+
+def prettyfypath(path):
```
Just for the sake of clarity, should this be _prettifypath_? :)

jerneju replied:
Yes, probably.

jerneju commented:
@ajdapretnar : Well, **pdfminer3k** may be a little better module to extract texts from _pdf_ format files. However that module does not work on all required platforms.

kernc commented:
What about [pdfrw](https://github.com/pmaupin/pdfrw) (seems recent) or other [pure-python alternatives](https://github.com/pmaupin/pdfrw#pure-python)?

jerneju commented:
@kernc: Thank you, now I tested by myself. **pdfminer3k** can be installed on Windows.

kernc commented:
Can we also support epub? https://github.com/aerkalov/ebooklib

kernc commented:
Not _epub_! Its docs link to some page in French. ebooklib.

kernc commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,217 @@
+        super().__init__(path)
+
+    def read_file(self):
+        fp = open(self.path, 'rb')
```
with open.

jerneju replied:
Fixed.

kernc commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,217 @@
+            for lt_obj in layout:
+                if isinstance(lt_obj, LTTextBox) or isinstance(lt_obj, LTTextLine):
+                    extracted_text += lt_obj.get_text()
+        self.content = extracted_text
```
make
```py
text = []
for page in pages:
   text.append(page.get_text())
text = ' '.join(text)
```

kernc commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,217 @@
+    def read_file(self):
+        odtfile = load(self.path)
+        texts = odtfile.getElementsByType(text.P)
+        self.content = " ".join(teletype.extractText(texts[i]) for i in range(len(texts)))
```
```py
' '.join(teletype.extractText(t) for t in texts)
```

jerneju replied:
Yes, sure.

kernc commented:
Add .html/.xml support via beautifulsoup (`root_node.getText()`) which is already a dependency.

ajdapretnar commented:
Ok, how about we stop supporting all possible formats and focus on stabilizing and testing this widget? We need it soon and test it thoroughly before the workshop, so as we said: support basic formats and have a solid widget and later add support for n formats. Thanks!

ajdapretnar commented:
Works fine for xml, too. For now I see no problems with the widget, but it needs thorough testing. Could you add good tests for the widget, please?

Also, I am seeing this for some pdf files.
<img width="618" alt="screen shot 2017-07-24 at 16 07 50" src="https://user-images.githubusercontent.com/12524972/28529981-7f4ee72c-7092-11e7-8591-d33fc38e5c35.png">

Is this a bug in the reader or something we can fix?

jerneju commented:
That is a bug in the reader. I can test this document using another pdf library.

jerneju commented:
Fixed.

ajdapretnar commented:
There are a couple of issues still pertaining.
1. Widget crashes with I'm loading some data and pressing Cancel in the middle of it. It breaks the entire Python, as a matter of fact.
2. Set _content_ to be a text_feature, not name. You need to set it with set_text_features().
3. When I load Zdravljica files, they seem to render ok in Corpus Viewer, while I am only getting the first line in Data Table. Any idea why this happens?
4. The output of the widget should be Corpus, not Data. Just a matter of renaming it.

ajdapretnar commented:
5. A crash occurs when connecting it to Scatter Plot. Probably an issue of SP itself, but just mentioning here, if perhaps you can check what's the problem.

Traceback (most recent call last):
  File "/Users/ajda/orange/orange3/Orange/canvas/scheme/widgetsscheme.py", line 832, in process_signals_for_widget
    widget.handleNewSignals()
  File "/Users/ajda/orange/orange3/Orange/widgets/visualize/owscatterplot.py", line 404, in handleNewSignals
    self.sparse_to_dense(self.subset_data))
  File "/Users/ajda/orange/orange3/Orange/widgets/visualize/owscatterplotgraph.py", line 609, in new_data
    self.set_data(data, **args)
  File "/Users/ajda/orange/orange3/Orange/widgets/utils/scaling.py", line 100, in set_data
    self._compute_scaled_data()
  File "/Users/ajda/orange/orange3/Orange/widgets/utils/scaling.py", line 52, in _compute_scaled_data
    self.valid_data_array = ~np.isnan(no_jit)
TypeError: ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

nikicc commented:
The above error sounds like a problem with sparse matrices. Are we outputting sparse matrices from Text Import?

jerneju commented:
@ajdapretnar @nikicc : Scatter Plot fails on empty sparse X matrix. Yes, corpus created by this widget has empty sparse X. It is automatically created.

nikicc commented:
@jerneju could we then fix this in Scatter Plot?

jerneju commented:
@ajdapretnar: Thank you for finding new errors. Let's go:
1) Maybe Import Images crashes as well? I shall check this.
2) Fixed.
3) Because of newlines. I can replace them with spaces. Only first line is seen in Table.
4) Fixed.

ajdapretnar commented:
I think we should. Some folks might find it natural to connect Scatter Plot to Text Import and getting an error is a bad surprise. 😁

jerneju commented:
@nikicc : Yes, I am planning to fix Scatter Plot.

ajdapretnar commented:
> Because of newlines. I can replace them with spaces. Only first line is seen in Table.

Hm. I'm not sure. What do you think @nikicc? I like poem for example to be nicely formatted in Corpus Viewer... If it's a poem, it should look like a poem. But it does look weird in a Data Table. Opinions?

nikicc commented:
> Because of newlines. I can replace them with spaces. Only first line is seen in Table.

I'm not sure about this. In Corpus Viewer it probably looks much better with the newlines? Table is not meant to be a widget for viewing corpuses anyhow.

nikicc commented:
> Hm. I'm not sure. What do you think @nikicc? I like poem for example to be nicely formatted in Corpus Viewer... If it's a poem, it should look like a poem. But it does look weird in a Data Table. Opinions?

I Would keep newlines for the sake of Corpus Viewer. Maybe Table's view component could be hacked to remove newlines?

ajdapretnar commented:
True. Then I guess let's go with newlines and embrace them fully. If there's a newline in a text, we should keep it.

ajdapretnar commented:
p.s. How does tokenization handle newlines?

jerneju commented:
Scatter Plot issue fixed.

nikicc commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+            categories.append(textdata.category)
+        return list(set(categories))
+
+def scan(topdir, include_patterns=("*",), exclude_patterns=(".*",)):
```
How do I pass `include_patterns ` or `exclude_patterns ` from `TextScan` class; i.e. how can the user scan directories and load only `txt` for example?

I this is not possible, why the fuss about this?

jerneju replied:
Fixed.

nikicc commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+    def read(self, replace_white_space=False):
+        try:
+            textdata = self.read_file()
+        except:
```
The same happens for [webtext corpus](https://github.com/nltk/nltk_data/blob/gh-pages/packages/corpora/webtext.zip). Out of 6 text files, I only got two if TxtReader is in text mode.

ajdapretnar commented on orangecontrib/text/widgets/owimporttexts.py:
```diff
@@ -0,0 +1,642 @@
+
+class OWImportTexts(widget.OWWidget):
+    name = "Import Texts"
+    description = "Import text document(s) from a directory(s)"
```
I also vote for: Import text documents from folders.
Plural is enough and perhaps folder is more intuitive to directory?

jerneju replied:
Fixed.

ajdapretnar commented on orangecontrib/text/widgets/owimporttexts.py:
```diff
@@ -0,0 +1,642 @@
+        """
+        corpus = None
+        if self._text_data:
+            names = ["name", "path", "content"]
```
So TextScan retrives documents from a folder and builds a corpus? Perhaps a rename is in order then? Sth like RetrieveText or GetText? But I agree, scripting should be supported, too.

jerneju replied:
Scripting is now supported. Now we need to choose a proper name.

- `TextScan` (current)
- `RetrieveText`
- `GetText`

jerneju replied:
`ImportDocuments` Same name for widget and scripting part.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+        for textdata in text_data:
+            categories.append(textdata.category)
+        return list(set(categories))
+
```
Fixed.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+
+    @classmethod
+    def get_categories(cls, text_data):
+        categories = []
```
Fixed.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+                batch = []
+
+        if batch and self.report_progress is not None:
+            self.report_progress(
```
Fixed.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+        for path in filescanner:
+            if fnmatch_any(path, patterns):
+                reader = Reader.get_reader(path)
+                textdata = reader.read()
```
Sure.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+        with open(self.path, 'rb') as f:
+            content = f.read()
+        self.content = content
+        return self.make_text_data()
```
Yes, fixed.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+
+    def read_file(self):
+        with open(self.path, 'rb') as f:
+            content = f.read()
```
Fixed.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+            textdata = self.read_file()
+        except:
+            textdata = None
+        self.replace_white_space=replace_white_space
```
Fixed.

jerneju commented on orangecontrib/text/widgets/owimporttexts.py:
```diff
@@ -0,0 +1,642 @@
+    name = "Import Texts"
+    description = "Import text document(s) from a directory(s)"
+    icon = ""
+    priority = 110
```
Fixed.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+TextDataError.isvalid = property(lambda self: False)
+
+
+class Reader():
```
Fixed.

jerneju commented on orangecontrib/text/widgets/owimporttexts.py:
```diff
@@ -0,0 +1,642 @@
+from Orange.widgets.widget import Output
+
+from orangecontrib.text.corpus import Corpus
+from orangecontrib.text.widgets.utils.readers import TextScan
```
All right, I am moving there. But are you sure that `text_import.py` is a proper file name? Maybe a name without an underscore?

jerneju replied:
`import_documents.py`

nikicc commented:
@jerneju the icon for this is in #289.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+    ext = ".txt"
+
+    def read_file(self):
+        with open(self.path, 'rb') as f:
```
`detect_encoding` from _Orange.data.io_

ajdapretnar commented:
I get an error when loading my articles:
Traceback (most recent call last):
  File "/Users/ajda/orange/orange3-text/orangecontrib/text/widgets/owimportdocuments.py", line 537, in __onRunFinished
    getattr(self.Warning, error)()
  File "/Users/ajda/orange/orange3/Orange/widgets/utils/messages.py", line 114, in __call__
    self.formatted = self.format(*args, **kwargs)
KeyError: "'U'"

ajdapretnar commented:
Cancel still doesn't work for me.
Traceback (most recent call last):
  File "/Users/ajda/orange/orange3-text/orangecontrib/text/widgets/owimportdocuments.py", line 547, in cancel
    assert self.__pendingTask is not None
AssertionError

jerneju commented:
@ajdapretnar : First error is now probably fixed.

ajdapretnar commented:
In readin xml files, I am getting:
Traceback (most recent call last):
  File "/Users/ajda/orange/orange3-text/orangecontrib/text/import_documents.py", line 62, in read
    self.read_file()
  File "/Users/ajda/orange/orange3-text/orangecontrib/text/import_documents.py", line 147, in read_file
    soup = BeautifulSoup(markup.read(), "lxml")
  File "/Users/ajda/anaconda/envs/o3/lib/python3.5/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x92 in position 16794: invalid start byte

This is however caught by a Warning I assume (607 problems during execution), which is good. However, I'm wondering whether we can fix this at source, since I guess it's just a decoding issue and can be resolved. :)

ajdapretnar commented:
I was waiting for some progress on the data, when I decided to quit Orange. I got this in the console:
Traceback (most recent call last):
  File "/Users/ajda/anaconda/envs/o3/lib/python3.5/logging/__init__.py", line 1881, in shutdown
    h.flush()
  File "/Users/ajda/anaconda/envs/o3/lib/python3.5/logging/__init__.py", line 964, in flush
    self.stream.flush()
  File "/Users/ajda/orange/orange3/Orange/canvas/application/outputview.py", line 207, in flush
    self.flushed.emit()
AttributeError: 'TextStream' does not have a signal with the signature flushed()

ajdapretnar commented:
Also, in the PR, I'm constantly seeing some messages in the console, whenever I'm placing or moving widgets around.
INFO:Orange.canvas.scheme.scheme:Added node 'Preprocess Text' to scheme ''.
INFO:Orange.canvas.canvas.scene:Added item '<Orange.canvas.canvas.items.nodeitem.NodeItem object at 0x188f39ca8>' to 'CanvasScene(objectName='', ...)'
INFO:Orange.canvas.scheme.scheme:Added link 'Import Documents' ('Corpus') -> 'Preprocess Text' ('Corpus') to scheme ''.
INFO:Orange.canvas.scheme.signalmanager:Link added (SchemeLink((Import Documents, Corpus) -> (Preprocess Text, Corpus))). Scheduling signal data update.
INFO:Orange.canvas.canvas.scene:Added link 'Import Documents' -> 'Preprocess Text' to 'CanvasScene(objectName='', ...)'
INFO:Orange.canvas.canvas.scene:Setting interaction 'None' to 'CanvasScene(objectName='', ...)'

I don't think this should be displayed in there. Is this some forgotten print somewhere?

jerneju commented:
@ajdapretnar : Please test all of them except last two. Those two issues occurs because of **pdfminer3k**.

kernc commented on orangecontrib/text/import_documents.py:
```diff
@@ -0,0 +1,262 @@
+)
+TextDataError.isvalid = property(lambda self: False)
+
+log = logging.getLogger()
```
Add `__name__`

jerneju replied:
Added.

kernc commented on orangecontrib/text/import_documents.py:
```diff
@@ -0,0 +1,262 @@
+
+
+class Reader:
+    _readers = ["TxtReader", "DocxReader", "OdtReader", "PdfReader", "XmlReader"]
```
You can make this:
```py
class Reader(metaclass=Orange.util.Registry):
```
The all reader implementations can be iterated over with:
```py
for reader in cls.registry:
    if ext in reader.ext:
        ...
```

jerneju replied:
I like that.

kernc commented on orangecontrib/text/import_documents.py:
```diff
@@ -0,0 +1,262 @@
+
+
+class XmlReader(Reader):
+    ext = [".xml"]
```
Just curious, why not html and newspaper?

jerneju replied:
I had some issues and `html` is not so high priority at the moment. But no worries, I am planning to implement other formats as well.

kernc commented on orangecontrib/text/import_documents.py:
```diff
@@ -0,0 +1,262 @@
+            if fnmatch.fnmatch(dirname, ".*"):
+                dirnames.remove(dirname)
+
+        def matches_any(fname, patterns):
```
Move definition outside of loop.

jerneju replied:
Fixed.

jerneju commented on orangecontrib/text/widgets/utils/readers.py:
```diff
@@ -0,0 +1,215 @@
+        return self.make_text_data()
+
+
+DefaultFormats = ("docx", "odt", "txt", "pdf", "xml")
```
Moved.

nikicc commented:
I only found one small glitch. If I click cancel, the Info box says `cancelled` but old corpus is still on the output. It probably shouldn't be anymore.

nikicc commented:
Shit, pdfminer3k is not available on conda, at least for linux-64 channel ([see this PR](https://github.com/conda-forge/orange3-text-feedstock/pull/4)). @jerneju is there something else we could use, or should we start working towards putting it on conda-forge?

@ajdapretnar we didn't plan to send installation instructions for conda right?

jerneju commented:
@nikicc: **PyPDF2** (https://anaconda.org/conda-forge/pypdf2) Firstly I implemented that module but then I moved to **pdfminer3k**. I think the latter is better.
